### PR TITLE
fix: got rid of the memory leak

### LINKED ISSUE #913: Notebooks: always show the vizualization options
It's a little jarring to have so much stuff show when you hover over a graph. should we just always show it?

image

### LINKED ISSUE #928: Simple Table should key off result name + table number
today, simple table just keys off table number, which will group results together in a strange way. 

we should use result name + table number

### PR DESCRIPTION
**Pull request by drdelambre**
the ref context was causing a bunch of issues on components when they all worked together, but not when the components were used in isolation. this was only being used for the panel focus mechanism, so i removed it. seemed to line up with whatever russ wanted to do anyways.

also memoizes everything... cause i was in there

and adds the name of the yield to the table header, while improving how the tables were grouped / dedupped in simple table

closes #928 
closes #913

### CODE REVIEW

**Comment by hercules-influx** [comment]
Metric | Score | Level
------------ | ------------- | ------------- 
performance | 37 | :red_circle:
accessibility | 96 | :green_circle:
best-practices | 93 | :green_circle:
seo | 90 | :green_circle:
pwa | null | :question:


[Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?vars%5Blighthouse_score_branch%5D=958)

File: src/flows/context/query.tsx [hunk 1]
```diff
@@ -255,12 +255,20 @@ export const QueryProvider: FC = ({children}) => {
...
       })
       .then(raw => {
-        return {
-          source: text,
-          raw: raw.csv,
-          parsed: fromFlux(raw.csv),
-          error: null,
-        } as FluxResult
+        return new Promise((resolve, reject) => {
+          requestAnimationFrame(() => {
```
Comment (drdelambre) [inline src/flows/context/query.tsx:259]: this was an attempt to offload some of the freezing that was happening on the page while it was parsing 4 csv files at once. it helped a little, but i think moving parsing into a service worker is the next step

File: src/flows/pipes/Downsample/view.tsx [hunk 2]
```diff
@@ -20,23 +20,39 @@ import DurationInput from 'src/shared/components/DurationInput'
...
+    (fn: string) => {
+      const fns = options.map(f => ({name: f}))
+      let found = false
+      let fnIdx = fns.findIndex(f => f.name === fn)
+
+      while (fnIdx !== -1) {
+        found = true
+        fns.splice(fnIdx, 1)
+        fnIdx = fns.findIndex(f => f.name === fn)
+      }
```
Comment (genehynson) [inline src/flows/pipes/Downsample/view.tsx:44]: wait so what's the point of this while loop? Seems like it should only run once?

Comment (genehynson) [reply src/flows/pipes/Downsample/view.tsx:44]: I'm nervous this will lead to an infinite loop

File: src/flows/pipes/Visualization/Controls.tsx [hunk 3]
```diff
@@ -42,42 +44,49 @@ const Controls: FC<Props> = ({toggle, visible}) => {
...
+    if (!data.functions || !data.functions.length) {
+      return [{name: 'mean'}].map(f => f.name)
     }
+    return data.functions.map(f => f.name)
+  }, [data.functions])
 
-    if (!found) {
-      fns.push({name: fn})
-    }
+  const selectFn = useCallback(
```
Comment (genehynson) [inline src/flows/pipes/Visualization/Controls.tsx:54]: any chance this duplicated code could be moved to a central spot?

File: src/flows/pipes/Visualization/view.tsx [hunk 4]
```diff
@@ -69,7 +69,7 @@ const Visualization: FC<PipeProp> = ({Context}) => {
         <div className="flow-visualization">
           <div className="flow-visualization--view">
             <View
-              loading={loading}
+              loading={RemoteDataState.Done}
```
Comment (genehynson) [inline src/flows/pipes/Visualization/view.tsx:72]: should this be `RemoteDataState.Loading`?

File: src/visualization/types/SimpleTable/PagedTable.tsx [hunk 5]
```diff
@@ -87,18 +96,21 @@ const subsetResult = (
 
   // group by table id (series)
   for (let ni = 0; ni < page; ni++) {
-    if (subset['table'].data[ni] === lastTable) {
+    if (
+      `y${subset['result'].data[ni]}:t${subset['table'].data[ni]}` === lastTable
```
Comment (genehynson) [inline src/visualization/types/SimpleTable/PagedTable.tsx:100]: nit: extra space

**Comment by genehynson** [comment]
> Metric	Score	Level
> performance	34	🔴
> accessibility	96	🟢
> best-practices	93	🟢
> seo	90	🟢
> pwa	null	❓
> [Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?vars%5Blighthouse_score_branch%5D=958)

Performance dropped to 34? 🤔  Maybe it was a flake - hopefully a re-run brings it back up

**Comment by Asalem1** [comment]
This one is crushing our lighthouse scores down 9. initiate public shaming

File: src/flows/pipes/Downsample/view.tsx (hunk 2 above)
Comment (drdelambre) [reply src/flows/pipes/Downsample/view.tsx:44]: its for when we have multiple instances of a function being populated into the functions list.. and that stuff gets persisted to local storage. it's like a bug that just wont go away.

do you have an irrational fear of loops?

File: src/flows/pipes/Visualization/Controls.tsx (hunk 3 above)
Comment (drdelambre) [reply src/flows/pipes/Visualization/Controls.tsx:54]: could be.. i know the code is copy pasta, but i'm not sure how tightly coupled the function selection is between them right now as the downsample version of this is.. brand new and looks weird.

File: src/flows/pipes/Visualization/view.tsx (hunk 4 above)
Comment (drdelambre) [reply src/flows/pipes/Visualization/view.tsx:72]: the View requires a loading state, and the parent handles that loading state spinner, removing the need for it to be handled by the View. The way it was, with both handling the issue, we got two spinners on a visualization

making it Loading here would mean the view is always behind a spinner.

**Comment by drdelambre** [comment]
those in glass houses @Asalem1 ....

**Comment by drdelambre** [comment]
@genehynson master is at 37 right now? seems like something else is going on?

File: src/flows/pipes/Visualization/view.tsx (hunk 4 above)
Comment (genehynson) [reply src/flows/pipes/Visualization/view.tsx:72]: ah okay, got it so the intention is to remove the duplicate spinner 👍

File: src/flows/pipes/Downsample/view.tsx (hunk 2 above)
Comment (genehynson) [reply src/flows/pipes/Downsample/view.tsx:44]: So like if `fn` isn't in `fns`, `fnIdx` will always be `-1` right? Are you saying that there is no chance of that happening? 

Yes I had a tragic early career experience regarding infinite loops that led to infinite-loop-ptsd 😆

**Comment by genehynson** [comment]
> @genehynson master is at 37 right now? seems like something else is going on?

Looking at the dashboard for this branch it seems like the last 3 runs on this branch produced a 34 or 37 score tho 🤔

File: src/flows/pipes/Downsample/view.tsx (hunk 2 above)
Comment (drdelambre) [reply src/flows/pipes/Downsample/view.tsx:44]: yes

it's gonna be alright buddy.. embrace the while loop AND the for loop. no function stack creation overhead when iterating on large data sets? yes PLEASE

**Comment by genehynson** [comment]
> Metric	Score	Level
> performance	44	🔴
> accessibility	96	🟢
> best-practices	93	🟢
> seo	90	🟢
> pwa	null	❓
> [Compare your scores with master branch](https://influxdb.aws.influxdata.io/orgs/844910ece80be8bc/dashboards/073c4b95f1089000?vars%5Blighthouse_score_branch%5D=958)

back up to 44 - nice

File: src/flows/pipes/Downsample/view.tsx (hunk 2 above)
Comment (genehynson) [reply src/flows/pipes/Downsample/view.tsx:44]: alright I got the boatie guarantee or my money back haha

**Comment by drdelambre** [comment]
any comments about the PR or....?

**Review by genehynson — APPROVED** [review]

**Pull request merged**